### DISCUSSION TITLE
Next.js 13 - How to get params from url using userParams and statically pre render components at built time at same time

### DISCUSSION THREAD

**Discussion by umairjameelprixite** [vercel/next.js#51344] [Help]
### Summary

I am learning latest nextjs 13.4 and I have created a page on route /notes/[note]. Now I want to get the note param from url which I am doing using import { useParams } from 'next/navigation'.

Along with this, I want to statically Pre Render data for this route, so when I build the project, this component get its html file generated in .next folder. For this, I am setting up pre-defined note ids using [generateStaticParams](https://nextjs.org/docs/app/api-reference/functions/generate-static-params) (it works like getStaticPaths in previous next.js versions).

Now the problem is that, when I leave it as server component, useParams doesn't work and throws an error and when I set it a client component using 'user client' at top, it gets compiled but Static Pre-rendering for given 4 note ids doesn't work.

How to make both the things work for this component?

When I remove 'use client' from top and remove userParams from file. When I build project, 4 html files successfully gets generated as shown in image below:
<img width="320" alt="Screenshot 2023-06-15 at 3 25 40 PM" src="https://github.com/vercel/next.js/assets/101325439/ea64ebad-2137-41d1-a46d-28b09acda035">

### Additional information

```js
'use client'
import styles from '../../page.module.css'
import { useParams } from 'next/navigation'

async function getNoteData(id : string) {
  const note = await fetch(`https://jsonplaceholder.typicode.com/todos/${id}`);
  if(!note) {
    throw new Error("No data found.");
  }
  return note.json();
}

interface Note {
  userId: string;
  id: number;
  title: string;
  completed: string;
}

export function generateStaticParams() {
  return [{note: '1'}, {note: '2'}, {note: '3'}, {note: '4'}]
}

export default async function Note({params}: {params: {note: string}}) {
  const param = useParams();
  const todo : Note = await getNoteData(param.note);
  return (
    <main className={styles.main}>
      <h1>Note Page</h1>
      <div className={styles.description}>
          <p>{todo?.title}</p>
      </div>
    </main>
  )
}
```

**Comment by icyJoseph** [comment]
Why not use the note from `params`?

```tsx
// "use client" // remove this

export default async function Note({params}: {params: {note: string}}) {
  const todo : Note = await getNoteData(params.note);
  return (
    <main className={styles.main}>
      <h1>Note Page</h1>
      <div className={styles.description}>
          <p>{todo?.title}</p>
      </div>
    </main>
  )
}
```

It contains the info you need.

Otherwise, if you really want to use the client hook, for some reason, refactor. Though it is best to get the data in the server component.

```tsx
'use client'

// src/app/notes/[note]/note-client.tsx
export const ClientNote = ({note}: {note: Note}) => {
    const param = useParams();
    return <div>Something</div>
}
```

And then in the page component:

```tsx
// src/app/notes/[note]/page.tsx
import { ClientNote } from './note-client';

// rest of the file

export default async function Note({params}: {params: {note: string}}) {
  const todo : Note = await getNoteData(param.note);
  return (
    <main className={styles.main}>
      <h1>Note Page</h1>
      <div className={styles.description}>
          <p>{todo?.title}</p>
      </div>

     <ClientNote note={todo} />
    </main>
  )
}
```

The root issue is that client components can't be async. And so you have to structure your code accordingly. A client component, rendered by a server component, where the client component uses the params client hook.

**Reply by umairjameelprixite** [reply]
@icyJoseph Thanks for the answer.
Well, I am returning only 4 note ids from generateStaticParams but there can be more than 4 note ids comes from url on client side. That's why I want to fetch it from url. 

Well, as per your solution, I have created another client component where I can fetch id of note from url. So, in this component, I will have to call api again with new id that I have fetched from url?

I was thinking of using same function getNoteData() for both pre-rendering and also for fetch new note data for different ids coming from url.

**Reply by icyJoseph** [reply]
> Well, I am returning only 4 note ids from generateStaticParams but there can be more than 4 note ids comes from url on client side. That's why I want to fetch it from url.

In the old Pages directory, you'd return a flag called `fallback: true` or `fallback: "blocking"`, and that'd allow for pages not pre-rendered at build time, to be generated at runtime,. 

AFAIK, in the app dir, that's like, by default, if the note doesn't exist, you can call `notFound`, and that ops out. 

- https://nextjs.org/docs/app/api-reference/functions/not-found

And that should be all you need really. I am a bit busy so maybe I missed something.

```tsx
import { notFound } from "next/navigation";

async function getNoteData(id: string) {
  if (Number(id) > 5) notFound();

  const note = await fetch(`https://jsonplaceholder.typicode.com/todos/${id}`);
  if (!note) {
    notFound();
  }
  return note.json();
}

interface Note {
  userId: string;
  id: number;
  title: string;
  completed: string;
}

export function generateStaticParams() {
  return [{ note: "1" }, { note: "2" }, { note: "3" }, { note: "4" }];
}

export default async function Note({ params }: { params: { note: string } }) {
  const todo: Note = await getNoteData(params.note);

  return (
    <main>
      <h1>Note Page</h1>
      <div>
        <p>{todo?.title}</p>
      </div>
    </main>
  );
}

```

So you'd something like this, and then you could render note 5, which is not part of the initial set, but nothing more than 6.
